File: Dinotopiada soyuq günlər.py
# https://basecamp.eolymp.com/az/problems/9641
import math

class Cord:
    def __init__(self, x, y):
        self.x = x
        self.y = y

def cmpx(b, c):
    if b.x != c.x:
        return b.x < c.x
    return b.y < c.y


def get_dist(b, c):
    return (b.x - c.x) ** 2 + (b.y - c.y) ** 2


def cmpd(b, c, zro):
    return get_dist(zro, b) < get_dist(zro, c)

def main():
    n = int(input())  
    a = [] 
    for _ in range(n):
        x, y = map(int, input().split())
        a.append(Cord(x, y)) 

    zro = Cord(0, 0) 

    a.sort(key=lambda p: (p.x, p.y))

    ans = float('inf') 

    for i in range(1, n):
        ans = min(ans, get_dist(a[i-1], a[i]))

    m = 100
    for i in range(n):
        for j in range(max(0, i - m), min(n, i + m + 1)):
            if i == j:
                continue
            ans = min(ans, get_dist(a[i], a[j]))

    a.sort(key=lambda p: get_dist(zro, p))
   
    for i in range(1, n):
        ans = min(ans, get_dist(a[i-1], a[i]))

    print(ans) 

if __name__ == "__main__":
    main()

File: Dinotopiada soyuq günlər.cpp
// https://basecamp.eolymp.com/az/problems/9641
#include<bits/stdc++.h>
using namespace std;
int n;
struct cord{
	int x, y;
} a[100010];

bool cmpx(cord b, cord c){
	if(b.x != c.x) return b.x < c.x;
	return b.y < c.y;
}

long long get_dist(cord b, cord c){
	return 1ll * (b.x - c.x) * (b.x - c.x) + 1ll * (b.y - c.y) * (b.y - c.y);
}

cord zro;

bool cmpd(cord b, cord c){
	return get_dist(zro, b) < get_dist(zro, c);
}

long long ans = (long long)1e18 + 7;

signed main(){
	ios_base::sync_with_stdio(0), cin.tie(0), cout.tie(0);
	cin >> n;
	for(register int i = 1; i <= n; ++i){
		cin >> a[i].x >> a[i].y;
	}
	const int m = 100;
	sort(a+1,a+n+1,cmpx);
	for(register int i = 1; i < n; ++i){
		ans = min(ans, get_dist(a[i], a[i+1]));
	}
	for(register int i = 1; i <= n; ++i){
		for(register int j = max(1, i - m); j <= min(n, i + m); ++j){
			if(i == j) continue;
			ans = min(ans, get_dist(a[i], a[j]));
		}
	}
	sort(a+1,a+n+1,cmpd);
	for(register int i = 1; i < n; ++i){
		ans = min(ans, get_dist(a[i], a[i+1]));
    }
	cout << ans << '\n';
	return 0;
}
